<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.1.189">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>additional_information</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
</style>


<script src="additional_information_files/libs/clipboard/clipboard.min.js"></script>
<script src="additional_information_files/libs/quarto-html/quarto.js"></script>
<script src="additional_information_files/libs/quarto-html/popper.min.js"></script>
<script src="additional_information_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="additional_information_files/libs/quarto-html/anchor.min.js"></script>
<link href="additional_information_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="additional_information_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="additional_information_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="additional_information_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="additional_information_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">



<section id="metashiny-app-allows-users-to-analyze-and-visualize-metagenomic-short-read-sequencing-data." class="level3">
<h3 class="anchored" data-anchor-id="metashiny-app-allows-users-to-analyze-and-visualize-metagenomic-short-read-sequencing-data.">MetaShiny app allows users to analyze and visualize metagenomic short-read sequencing data.</h3>
<ul>
<li><p>It takes a list containing the sample name and the condition.</p></li>
<li><p>The Fastq file must have a the same name as the sample name in the list.</p></li>
<li><p>The app will perform end to end analysis using the sequencing data with appropriate tools.</p></li>
</ul>
</section>
<section id="instructions" class="level1">
<h1>Instructions</h1>
<p>Code can be found on github: https://github.com/Nirmal2310/MetaShiny</p>
<p>Please post <a href="https://github.com/Nirmal2310/MetaShiny">issues on github</a>, and feel free to contribute by forking and submitting development branches.</p>
<p>To run this app locally on your machine, download R or RStudio in the local system.</p>
<p>You may now run the shiny app with just one command in R:</p>
<pre><code>shiny::runGitHub("MetaShiny", "Nirmal2310")</code></pre>
<p>If you are running the App for the first time please tick the checkbox “setup” in the Input Data tab to install the required tools, R packages and databases.</p>
</section>
<section id="input-data" class="level1">
<h1>Input Data</h1>
<p><a name="inputdata" href=""></a></p>
<section id="input-data-1" class="level2">
<h2 class="anchored" data-anchor-id="input-data-1">Input Data</h2>
<p>You can use this app by</p>
<ol type="1">
<li>Exploring the pre-loaded example data set. This is a pre-loaded metagenome DNA sequencing of ten samples example for exploring the app’s features.</li>
<li>Upload your own data</li>
</ol>
<p><a name="dataformat" href=""></a></p>
</section>
<section id="data-format" class="level2">
<h2 class="anchored" data-anchor-id="data-format">Data Format</h2>
<ul>
<li>Must be a .CSV <em>comma-separated-value</em> file.</li>
<li>File must have a header row.</li>
<li>First column must be named as <code>Sample_Id</code>.</li>
<li>Second column must be named as <code>Group</code>.</li>
</ul>
<p><img src="input_data.png" alt="Input Data" style="width: 25%"></p>
<p><a name="outputdata" href=""></a></p>
</section>
<section id="ouput-metadata" class="level2">
<h2 class="anchored" data-anchor-id="ouput-metadata"><strong>Ouput MetaData</strong></h2>
<ul>
<li>Each row of the 3rd column represents the ARG term.<br>
</li>
<li>Additional columns provide information about</li>
</ul>
<ol type="1">
<li>Sample Id</li>
<li>Bacterial Classification</li>
<li>ARO Term</li>
<li>Counts</li>
<li>ARG Length</li>
<li>Percentage Identity</li>
<li>Drug Class</li>
<li>Resistance Mechanism</li>
<li>AMR Gene Family</li>
<li>Percentage Coverage</li>
<li>Normalized Counts</li>
<li>Bacterial Family</li>
<li>Group</li>
</ol>
<p>All the subsequent visualizations will be done using this metadata.</p>
<p>Example file: &lt;Sample_information.csv&gt;</p>
<p>Analysis: When the list is uploaded, the data is then analyzed by the app. The app first utilizes fastp and bbtools for data pre-processing. The preprocessed FASTQ reads are then assembled using SPADES. The de novo assembled metagenome is then binned in MAGs using metaWRAP. The binned MAGs are annotated using GTDBtk. Antimicrobial Resistance Genes (AMRs) are then identified using RGI with CARD database. The counts for each ARG are calculated using SAMtools and the table from the unix pipeline contains the ARO term, Counts, Bacterial Classification and other terms. The counts for each ARG terms are then normalized using GPCM method in R and then utilized for further analysis and visualization in R.</p>
<p><img src="gpcm_equation.png" alt="Output Data" style="width: 50%"></p>
</section>
<section id="analyzed-data" class="level2">
<h2 class="anchored" data-anchor-id="analyzed-data">Analyzed Data</h2>
<p><img src="output_data.png" alt="Output Data" style="width: 75%"></p>
<p>Example file: &lt;consolidated_data.csv&gt;</p>
<p><a name="vis" href=""></a></p>
</section>
</section>
<section id="visualizations" class="level1">
<h1>Visualizations</h1>
<section id="arg-cohort-analysis" class="level2">
<h2 class="anchored" data-anchor-id="arg-cohort-analysis">ARG Cohort Analysis</h2>
<p><a name="cohortanalysis" href=""></a></p>
</section>
<section id="drug-class" class="level2">
<h2 class="anchored" data-anchor-id="drug-class">Drug Class</h2>
<p>This plot represents the resistance genes for antimicrobial drugs classified by the CARD database. The Y-axis represents the Drug class and X-axis represents the sum of all the ARG abundance belonging to the drug class.</p>
<p><img src="Drug_class.png" alt="Drug Class vs. ARG Abundance" style="width: 100%"></p>
</section>
<section id="resistance-mechanism" class="level2">
<h2 class="anchored" data-anchor-id="resistance-mechanism">Resistance Mechanism</h2>
<p>This plot displays the mechanism of resistance of each ARG, depicted as a proportion of all ARGs detected in the given cohort.</p>
<p><img src="Resistance_mechanism.png" alt="Resistance Mechanism" style="width: 100%"></p>
<p><a name="distributionplot" href=""></a></p>
</section>
</section>
<section id="arg-distribution-among-bacterial-species" class="level1">
<h1>ARG Distribution among Bacterial Species</h1>
<section id="arg-richness-per-bacterial-species" class="level2">
<h2 class="anchored" data-anchor-id="arg-richness-per-bacterial-species">ARG Richness Per Bacterial Species</h2>
<p>The circular plot shows the ARG richness (number of unique ARGs) for each bacterial species. The bars are grouped by the bacterial species family information. This plot shows the diversity of ARGs per bacterial species across the cohort.</p>
<p><img src="ARG_cohort_richness.png" alt="Cohort ARG Richness" style="width: 100%"></p>
</section>
<section id="arg-abundance-per-bacterial-species" class="level2">
<h2 class="anchored" data-anchor-id="arg-abundance-per-bacterial-species">ARG Abundance Per Bacterial Species</h2>
<p>The circular plot shows the AMR Gene Family abundance {(Normalized Counts/sum(Normalized counts))*100} that are grouped by the bacterial species. This plot shows the abundance of ARGs per bacterial species across the cohort.</p>
<p><img src="ARG_Cohort_Abundance.png" alt="Cohort ARG Abundance" style="width: 100%"></p>
<p>Both these plots can be utilized to target the most significant AMR causing bacterial species.</p>
<p><a name="alphadiversity" href=""></a></p>
</section>
<section id="alpha-diversity" class="level2">
<h2 class="anchored" data-anchor-id="alpha-diversity">Alpha Diversity</h2>
<p>The plot shows the alpha diversity for each ARG terms. Alpha diversity is a measure of the number of species that are present in a given community. This plot compares the ARG diversity between Control and Case.</p>
<p><img src="Alpha_diversity.png" alt="Alpha Diversity" style="width: 100%"></p>
</section>
<section id="abundance-diversity" class="level2">
<h2 class="anchored" data-anchor-id="abundance-diversity">Abundance Diversity</h2>
<p>The plot shows the comparison between the abundance of each ARG term between the Control and Case. This plot is useful to identify the ARGs that are abundant in the Control but not in the Case. The <em>p-value</em> was calculated using Kruskal-Walis test.</p>
<p><img src="Abundance_Kruskal_Walis.png" alt="Abundance Diversity" style="width: 100%"></p>
<p><a name="betadiversity" href=""></a></p>
</section>
<section id="pca-plot" class="level2">
<h2 class="anchored" data-anchor-id="pca-plot">PCA Plot</h2>
<p>This plot uses Principal Component Analysis (PCA) to calculate the principal components of the count data using data from all ARO terms. Samples are projected on the first two principal components (PCs) and the percent variance explained by those PCs are displayed along the x and y axes. Ideally your samples will cluster by group identifier.</p>
<p><img src="PCA_plot.png" alt="PCA Plot" style="width: 100%"></p>
</section>
<section id="heatmap" class="level2">
<h2 class="anchored" data-anchor-id="heatmap">HeatMap</h2>
<p>This plot shows Heatmap with ARG abundance in terms of the log2(Normalized Counts) across the samples. Both samples and ARGs were clustered based on ARG abundance with Euclidean distance by complete linkage hierarchical clustering.</p>
<p><img src="HeatMap.png" alt="HeatMap" style="width: 100%"></p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>